File: Chatting_Application_BE/src/main/java/vn/nphuy/chatapp/domain/Channel.java
package vn.nphuy.chatapp.domain;

import org.hibernate.annotations.Filter;
import org.hibernate.annotations.FilterDef;
import org.hibernate.annotations.SQLDelete;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import vn.nphuy.chatapp.util.constant.ChannelTypeEnum;

@Entity
@Table(name = "channels")
@SQLDelete(sql = "UPDATE channels SET deleted = true WHERE id=?")
@FilterDef(name = "deletedChannelsFilter")
@Filter(name = "deletedChannelsFilter", condition = "deleted = false")
@Getter
@Setter
public class Channel extends AbstractEntity {

  private String name;

  @Enumerated(EnumType.STRING)
  private ChannelTypeEnum type = ChannelTypeEnum.TEXT;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "server_id")
  private Server server;

  public Server getServer() {
    return server != null && !server.isDeleted() ? server : null;
  }

}

File: ChatSync_Application_FE/types/channel.ts
import { IBaseModel, IMember, IServer } from '.';

export enum ChannelTypeEnum {
  TEXT = 'TEXT',
}

export interface IChannel extends IBaseModel {
  name: string;
  type: ChannelTypeEnum;
  server: IServer;
}

export interface IMessage extends IBaseModel {
  content?: string;
  fileUrls?: string[];
  channel: IChannel;
  sender: IMember;
  deleted: boolean;
}
